File: components/backend/template-download/template_download.js
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const http = require('http');
const https = require('https');
const { spawnSync } = require('child_process');
const { readRecents } = require('../project-store');

const TEMPLATES_PATH = path.join(__dirname, '..', '..', 'templates.json');

function readTemplates() {
  try {
    if (!fs.existsSync(TEMPLATES_PATH)) return {};
    const raw = fs.readFileSync(TEMPLATES_PATH, 'utf8');
    const parsed = JSON.parse(raw);
    if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) {
      return parsed;
    }
  } catch (e) {
    console.error('Failed to read template definitions:', e);
  }
  return {};
}

function getImportTargetFolder() {
  const recents = readRecents() || {};
  const openFolder = typeof recents.open === 'string' && recents.open ? recents.open : null;
  if (!openFolder) {
    throw new Error('No open folder is available in recents.json.');
  }
  return openFolder;
}

function getTemplateDefinition(templateKey) {
  const templates = readTemplates();
  return templates[templateKey] || null;
}

function downloadFile(url, destPath) {
  return new Promise((resolve, reject) => {
    const transport = url.startsWith('https:') ? https : http;

    const request = transport.get(url, (response) => {
      if (response.statusCode && response.statusCode >= 300 && response.statusCode < 400 && response.headers.location) {
        response.resume();
        downloadFile(response.headers.location, destPath).then(resolve).catch(reject);
        return;
      }

      if (response.statusCode !== 200) {
        response.resume();
        reject(new Error(`Download failed with status ${response.statusCode}`));
        return;
      }

      const fileStream = fs.createWriteStream(destPath);
      response.pipe(fileStream);
      fileStream.on('finish', () => fileStream.close(resolve));
      fileStream.on('error', reject);
    });

    request.on('error', reject);
  });
}

function extractArchive(archivePath, extractDir) {
  fs.mkdirSync(extractDir, { recursive: true });
  const isWindows = process.platform === 'win32';

  if (isWindows) {
    const result = spawnSync(
      'powershell',
      [
        '-NoProfile',
        '-Command',
        `Expand-Archive -Path '${archivePath.replace(/'/g, "''")}' -DestinationPath '${extractDir.replace(/'/g, "''")}' -Force`,
      ],
      { encoding: 'utf8' }
    );

    if (result.status !== 0) {
      throw new Error(result.stderr || 'Failed to extract archive with PowerShell.');
    }
    return;
  }

  const result = spawnSync('unzip', ['-o', archivePath, '-d', extractDir], { encoding: 'utf8' });
  if (result.status !== 0) {
    throw new Error(result.stderr || 'Failed to extract archive with unzip.');
  }
}

function copyDirectoryContents(sourceDir, targetDir) {
  fs.mkdirSync(targetDir, { recursive: true });
  const entries = fs.readdirSync(sourceDir, { withFileTypes: true });

  for (const entry of entries) {
    const sourcePath = path.join(sourceDir, entry.name);
    const targetPath = path.join(targetDir, entry.name);

    if (entry.isDirectory()) {
      copyDirectoryContents(sourcePath, targetPath);
    } else {
      fs.mkdirSync(path.dirname(targetPath), { recursive: true });
      fs.copyFileSync(sourcePath, targetPath);
    }
  }
}

function getExtractedSourceRoot(extractDir) {
  const entries = fs.readdirSync(extractDir, { withFileTypes: true })
    .filter((entry) => entry.name !== '.DS_Store');

  if (entries.length === 1 && entries[0].isDirectory()) {
    return path.join(extractDir, entries[0].name);
  }
  return extractDir;
}

async function importTemplate(templateKey, targetDir = null) {
  const template = getTemplateDefinition(templateKey);
  if (!template) {
    throw new Error(`Template "${templateKey}" was not found.`);
  }

  const downloadUrl = template.dl || template.url;
  if (!downloadUrl) {
    throw new Error(`Template "${templateKey}" does not define a download URL.`);
  }

  const destinationDir = targetDir || getImportTargetFolder();
  fs.mkdirSync(destinationDir, { recursive: true });

  const tempRoot = fs.mkdtempSync(path.join(os.tmpdir(), 'mama-template-'));
  const archivePath = path.join(tempRoot, `${template.name || templateKey}.zip`);
  const extractDir = path.join(tempRoot, 'extracted');

  try {
    await downloadFile(downloadUrl, archivePath);
    extractArchive(archivePath, extractDir);

    const sourceRoot = getExtractedSourceRoot(extractDir);
    copyDirectoryContents(sourceRoot, destinationDir);

    return {
      success: true,
      destinationDir,
      templateKey,
      templateName: template.name || templateKey,
    };
  } finally {
    fs.rmSync(tempRoot, { recursive: true, force: true });
  }
}

module.exports = {
  readTemplates,
  getImportTargetFolder,
  getTemplateDefinition,
  importTemplate,
};

File: components/backend/template_download/template_download.py
#!/usr/bin/env python3
"""
template_download.py — Python equivalent of template_download.js
Handles downloading and extracting project templates.
"""

import os
import json
import shutil
import tempfile
import urllib.request
import subprocess
import sys
import logging
from pathlib import Path

logger = logging.getLogger('mama.template_download')

TEMPLATES_PATH = Path(__file__).resolve().parent.parent.parent / 'templates.json'
RECENTS_PATH = Path(__file__).resolve().parent.parent.parent / 'recents.json'


def read_templates() -> dict:
    """Read template definitions from templates.json."""
    if not TEMPLATES_PATH.exists():
        return {}
    try:
        data = json.loads(TEMPLATES_PATH.read_text('utf-8'))
        if isinstance(data, dict):
            return data
    except Exception as e:
        logger.error('Failed to read templates: %s', e)
    return {}


def get_import_target_folder() -> str:
    """Get the target folder for template import from recents."""
    if RECENTS_PATH.exists():
        try:
            recents = json.loads(RECENTS_PATH.read_text('utf-8'))
            open_folder = recents.get('open')
            if open_folder:
                return open_folder
        except Exception:
            pass
    raise ValueError('No open folder is available in recents.json.')


def download_file(url: str, dest_path: str):
    """Download a file from URL to destination path."""
    urllib.request.urlretrieve(url, dest_path)


def extract_archive(archive_path: str, extract_dir: str):
    """Extract a zip archive to the given directory."""
    os.makedirs(extract_dir, exist_ok=True)

    if sys.platform == 'win32':
        # Use PowerShell on Windows
        result = subprocess.run(
            ['powershell', '-NoProfile', '-Command',
             f"Expand-Archive -Path '{archive_path}' -DestinationPath '{extract_dir}' -Force"],
            capture_output=True, text=True
        )
        if result.returncode != 0:
            raise RuntimeError(result.stderr or 'Failed to extract archive with PowerShell.')
    else:
        # Use unzip on Unix
        result = subprocess.run(
            ['unzip', '-o', archive_path, '-d', extract_dir],
            capture_output=True, text=True
        )
        if result.returncode != 0:
            raise RuntimeError(result.stderr or 'Failed to extract archive with unzip.')


def copy_directory_contents(source_dir: str, target_dir: str):
    """Copy all contents from source directory to target directory."""
    os.makedirs(target_dir, exist_ok=True)
    for item in os.listdir(source_dir):
        source_path = os.path.join(source_dir, item)
        target_path = os.path.join(target_dir, item)
        if os.path.isdir(source_path):
            shutil.copytree(source_path, target_path, dirs_exist_ok=True)
        else:
            shutil.copy2(source_path, target_path)


def get_extracted_source_root(extract_dir: str) -> str:
    """Get the actual source root after extraction (handles single-dir archives)."""
    entries = [e for e in os.listdir(extract_dir) if e != '.DS_Store']
    if len(entries) == 1:
        single_path = os.path.join(extract_dir, entries[0])
        if os.path.isdir(single_path):
            return single_path
    return extract_dir


def import_template(template_key: str, target_dir: str = None) -> dict:
    """Import a template by key, downloading and extracting it."""
    templates = read_templates()
    template = templates.get(template_key)
    if not template:
        raise ValueError(f'Template "{template_key}" was not found.')

    download_url = template.get('dl') or template.get('url')
    if not download_url:
        raise ValueError(f'Template "{template_key}" does not define a download URL.')

    destination_dir = target_dir or get_import_target_folder()
    os.makedirs(destination_dir, exist_ok=True)

    temp_dir = tempfile.mkdtemp(prefix='mama-template-')
    archive_path = os.path.join(temp_dir, f"{template.get('name', template_key)}.zip")
    extract_dir = os.path.join(temp_dir, 'extracted')

    try:
        download_file(download_url, archive_path)
        extract_archive(archive_path, extract_dir)

        source_root = get_extracted_source_root(extract_dir)
        copy_directory_contents(source_root, destination_dir)

        return {
            'success': True,
            'destinationDir': destination_dir,
            'templateKey': template_key,
            'templateName': template.get('name', template_key),
        }
    finally:
        shutil.rmtree(temp_dir, ignore_errors=True)